core/src/utils/sync/racy_lock.rs

```diff
@@ -0,0 +1,174 @@
+use alloc::boxed::Box;
+use core::{
+    fmt,
+    ops::Deref,
+    ptr,
+    sync::atomic::{AtomicPtr, Ordering},
+};
+
+/// Thread-safe, non-blocking, "first one wins" flavor of `once_cell::sync::OnceCell`
+/// with the same interface as `std::sync::LazyLock`.
+///
+/// The underlying implementation is based on `once_cell::sync::race::OnceBox` which relies on
+/// `core::atomic::AtomicPtr` to ensure that the data race results in a single successful
+/// write to the relevant pointer, namely the first write.
+/// See <https://github.com/matklad/once_cell/blob/v1.19.0/src/race.rs#L294>.
+///
+/// Performs lazy evaluation and can be used for statics.
+pub struct RacyLock<T, F = fn() -> T>
+where
+    F: Fn() -> T,
+{
+    inner: AtomicPtr<T>,
+    f: F,
+}
+
+impl<T, F> RacyLock<T, F>
+where
+    F: Fn() -> T,
+{
+    /// Creates a new lazy, racy value with the given initializing function.
+    pub const fn new(f: F) -> Self {
+        Self {
+            inner: AtomicPtr::new(ptr::null_mut()),
+            f,
+        }
+    }
+
+    /// Forces the evaluation of the locked value and returns a reference to
+    /// the result. This is equivalent to the `Deref` impl, but is explicit.
+    ///
+    /// This method will not block the calling thread if another initialization
+    /// routine is currently running.
+    pub fn force(this: &RacyLock<T, F>) -> &T {
+        let mut ptr = this.inner.load(Ordering::Acquire);
```

Comment:
nit: can we add a comment that this implementation is (almost) copy/paste of `OnceBox::get_or_try_init()`?